### PR TITLE
GreetingModelView renamed GreetingDTO, annotation updates, test updates

### PR DESCRIPTION
**Pull request by pbisby-uno**
**Demo Feedback Changes**
- GreetingModelView class refactored as GreetingDTO.
- Test updates to reflect DTO object name change.
- Explicit No Arg constructors removed and replaced with Lombok annotation.
- Various variable name changes for clarity.

**UPDATE - Requested Changes**
- All reviewer comments addressed / changes implemented
- Add integration tests.

**UPDATE - Requested changes**
- Add ControllerAdvice exception handler class
- Update custom exception
- Add Greeting view class
- Clean up build.gradle file

### CODE REVIEW

File: src/main/java/com/unosquare/cvgenerator/configuration/MapperConfig.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.modelmapper.ModelMapper;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class MapperConfig {
+
+    @Bean
+    public ModelMapper setModelMapper() {
```
Comment (escullion) [inline src/main/java/com/unosquare/cvgenerator/configuration/MapperConfig.java]: Suggestion: for bean setup, naming convention is usually just the name of the bean:
```suggestion
    public ModelMapper modelMapper() {
```

File: src/main/java/com/unosquare/cvgenerator/controller/HelloController.java [hunk 2]
```diff
@@ -17,8 +17,8 @@
 
     @GetMapping("/hello")
     public String sayHello(Model theModel) {
-        GreetingModelView greetingModelView = greetingService.findById(1);
-        theModel.addAttribute("greeting", greetingModelView);
+        GreetingDTO greetingDTO = greetingService.findById(1);
```
Comment (escullion) [inline src/main/java/com/unosquare/cvgenerator/controller/HelloController.java]: Future change: can be updated when adding the proper endpoints, but just making a reminder to remove hardcoded id:
```suggestion
        GreetingDTO greetingDTO = greetingService.findById(id);
```

File: src/main/java/com/unosquare/cvgenerator/model/entity/Greeting.java [hunk 3]
```diff
@@ -3,13 +3,15 @@
...
+import lombok.NoArgsConstructor;
 
 import javax.persistence.*;
 
 @Entity
 @Table(name = "hello")
 @Data
 @AllArgsConstructor
+@NoArgsConstructor
 @Builder
```
Comment (escullion) [inline src/main/java/com/unosquare/cvgenerator/model/entity/Greeting.java]: Suggestion - use only one of `@AllArgsConstructor` or `@Builder`, shouldn't need both

File: src/main/java/com/unosquare/cvgenerator/service/GreetingServiceImpl.java [hunk 4]
```diff
@@ -1,31 +1,28 @@
...
 
     private final HelloRepository helloRepository;
+    private final ModelMapper modelMapper;
 
     @Override
-    public GreetingModelView findById(int id) {
+    public GreetingDTO findById(int id) {
         Optional<Greeting> result = helloRepository.findById(id);
         if (!result.isPresent()) {
             throw new RuntimeException(String.format("Cannot find object with id %s", id));
```
Comment (escullion) [inline src/main/java/com/unosquare/cvgenerator/service/GreetingServiceImpl.java]: Future change: would be a good place to have a warning/error log

Comment (escullion) [inline src/main/java/com/unosquare/cvgenerator/service/GreetingServiceImpl.java]: Future change: Use more specific exception (or create custom exception) to better indicate "Not found"

**Review by escullion — COMMENTED** [review]
Looks good! Left some suggestions and things to consider as this code base grows

File: build.gradle [hunk 5]
```diff
@@ -22,14 +27,26 @@ repositories {
...
+	compile 'org.springframework.boot:spring-boot-starter-web'
 	compileOnly 'org.projectlombok:lombok'
 	developmentOnly 'org.springframework.boot:spring-boot-devtools'
 	runtimeOnly 'org.postgresql:postgresql'
 	annotationProcessor 'org.projectlombok:lombok'
-	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+	testCompile 'org.springframework.boot:spring-boot-starter-test'
+	integTestCompile("io.rest-assured:rest-assured:$restAssuredVersion")
 	compile group: 'org.modelmapper', name: 'modelmapper', version: '2.3.5'
+	compile("com.h2database:h2:1.4.199")
```
Comment (escullion) [inline build.gradle]: Just being nit-picky 😛 - some suggestions on how we can organization dependencies:
- organise dependencies together by type  - annotationProcessor, compile, integTestCompile, testCompile etc - makes it easier to find dependencies when you need to change them
- use consistent dependency style - some use compile group, brackets vs no brackets, double quotes vs single quotes etc which can be changed to make the others
- externalise version numbers to your `ext` section

File: src/integTest/java/com/unosquare/cvgenerator/AbstractIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.unosquare.cvgenerator;
+
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.transaction.annotation.Transactional;
+
+@Transactional
```
Comment (escullion) [inline src/integTest/java/com/unosquare/cvgenerator/AbstractIntegrationTest.java:10]: Documentation - maybe add a comment to say this should be used for any integration test which doesn't require a controller (service/repository tests)

File: src/integTest/java/com/unosquare/cvgenerator/AbstractWebIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.boot.web.server.LocalServerPort;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringRunner;
+
+/**
+ * Base class to derive concrete web test classes from.
+ */
+@ContextConfiguration
+@RunWith(SpringRunner.class)
```
Comment (escullion) [inline src/integTest/java/com/unosquare/cvgenerator/AbstractWebIntegrationTest.java:18]: Worth coming back to see if we need these 2 annotation in future - using a different spring boot version but usually `@SpringBootTest` is all thats needed

File: src/main/java/com/unosquare/cvgenerator/service/GreetingServiceImpl.java [hunk 8]
```diff
@@ -16,16 +17,17 @@
...
 
     @Override
-    public GreetingModelView findById(int id) {
+    public GreetingDTO findById(Integer id) {
         Optional<Greeting> result = helloRepository.findById(id);
         if (!result.isPresent()) {
-            throw new RuntimeException(String.format("Cannot find object with id %s", id));
+            String error = String.format("Cannot find object with id %s", id);
+            log.error(error);
+            throw new IdDoesNotExistException(error);
```
Comment (escullion) [inline src/main/java/com/unosquare/cvgenerator/service/GreetingServiceImpl.java]: Something that would be great to see in future would be a `@ControllerAdvice` exception handler - something we add in another ticket, allows you to handle your exceptions in the one place without worrying about handling exceptions in your business logic if you don't need to (e.g overuse of try catch/throws method signature)

**Review by escullion — APPROVED** [review]
Looks good! Left some further enhancements that can be addressed in another ticket

File: src/main/java/com/unosquare/cvgenerator/controller/HelloController.java [hunk 9]
```diff
@@ -15,10 +16,10 @@
 
     private final GreetingService greetingService;
 
-    @GetMapping("/hello")
-    public String sayHello(Model theModel) {
-        GreetingModelView greetingModelView = greetingService.findById(1);
-        theModel.addAttribute("greeting", greetingModelView);
+    @GetMapping("/hello/{id}")
+    public String sayHello(@PathVariable("id") Integer id, Model theModel) {
```
Comment (BrownMark) [inline src/main/java/com/unosquare/cvgenerator/controller/HelloController.java:23]: is theModel needed here in the signature?

File: src/main/java/com/unosquare/cvgenerator/service/GreetingServiceImpl.java (hunk 8 above)
Comment (BrownMark) [reply src/main/java/com/unosquare/cvgenerator/service/GreetingServiceImpl.java]: Agreed Controller advice for the IdDoesnotExistException would work well here.

File: src/main/resources/templates/helloworld.html [hunk 10]
```diff
@@ -5,7 +5,7 @@
 </head>
 
 <body>
-    <p th:text="${greeting.greeting}">placeholder</p>
+    <p th:text="${greetingDTO.greeting}">placeholder</p>
```
Comment (BrownMark) [inline src/main/resources/templates/helloworld.html]: Should really be passing view models to views.

**Review by BrownMark — CHANGES_REQUESTED** [review]
Looking mostly good, just some quick comments.

File: src/main/java/com/unosquare/cvgenerator/controller/HelloController.java (hunk 9 above)
Comment (BrownMark) [reply src/main/java/com/unosquare/cvgenerator/controller/HelloController.java:23]: ohhh I didn't see the code on line 22 - used for the SSR. Got it :)

**Review by BrownMark — APPROVED** [review]

**Pull request merged**